equinix/data_source_network_device.go

```diff
@@ -0,0 +1,759 @@
+package equinix
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/equinix/ne-go"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+var neDeviceStates = []string{ // Not sure if other states should be included
+	// ne.DeviceStateClusterSetUpInProgress,
+	// ne.DeviceStateDeprovisioned,
+	// ne.DeviceStateDeprovisioning,
+	// ne.DeviceStateFailed,
+	// ne.DeviceStateInitializing,
+	ne.DeviceStateProvisioned,
+	// ne.DeviceStateProvisioning,
+	// ne.DeviceStateWaitingClusterNodes,
+	// ne.DeviceStateWaitingPrimary,
+	// ne.DeviceStateWaitingSecondary,
+}
+
+func createDataSourceNetworkDeviceSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		neDeviceSchemaNames["UUID"]: {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: neDeviceDescriptions["UUID"],
+		},
+		neDeviceSchemaNames["Name"]: {
+			Type:     schema.TypeString,
+			Optional: true,
```

Comment:
We don't need optional flag any more in case of data source. I think for most of the schema defined here, we cannot reuse from resource_network_device.go.

Reply:
the 'optional' here is to permit the user to specify the UUID *or* the device name (but not both).  Right now this validated in the code itself (in `dataSourceNetworkDeviceRead()`).  I saw elsewhere that there is a schema attribute "ExactlyOneOf" that expresses this clearly, so I'll use this instead, as that will be more clear and explicit.